src/platforms/linux/linuxcryptosettings.cpp

```diff
@@ -3,15 +3,90 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 #include "cryptosettings.h"
+#include "logger.h"
 
-void CryptoSettings::resetKey() {}
+#include <QRandomGenerator>
+
+// No extra QT includes after this line!
+#undef Q_SIGNALS
+#include <libsecret/secret.h>
+
+const SecretSchema* cryptosettings_get_schema(void) {
+  static const SecretSchema cryptosettings_schema = {
+    "org.mozilla.vpn.cryptosettings", SECRET_SCHEMA_NONE, {
+      { "NULL", SECRET_SCHEMA_ATTRIBUTE_STRING },
+    }
+  };
+  return &cryptosettings_schema;
+}
+
+namespace {
+Logger logger("CryptoSettings");
+
+bool s_initialized = false;
+QByteArray s_key;
+}  // namespace
+
+
+void CryptoSettings::resetKey() {
+  logger.debug() << "Reset the key in the keychain";
+
+  GError* error = nullptr;
+  gboolean ok = secret_password_clear_sync(cryptosettings_get_schema(), nullptr,
+                                           &error, nullptr);
+  if (error != nullptr) {
+    logger.error() << "Key reset failed:" << error->message;
+    g_error_free(error);
+  }
+
+  s_initialized = false;
+}
 
 bool CryptoSettings::getKey(uint8_t key[CRYPTO_SETTINGS_KEY_SIZE]) {
-  Q_UNUSED(key);
+  if (!s_initialized) {
+    GError* error = nullptr;
+    gchar* password = secret_password_lookup_sync(cryptosettings_get_schema(),
+                                                  nullptr, &error, nullptr);
+    if (error != nullptr) {
+      logger.error() << "Key lookup failed:" << error->message;
+      g_error_free(error);
+      error = nullptr;
+    }
+
+    if (password != nullptr) {
+      QString b64key(password);
+      s_key = QByteArray::fromBase64(b64key.toUtf8());
+    } else {
+      logger.debug() << "Key not found. Let's create it.";
+      s_key = QByteArray(CRYPTO_SETTINGS_KEY_SIZE, 0x00);
+      QRandomGenerator* rg = QRandomGenerator::system();
+      for (int i = 0; i < CRYPTO_SETTINGS_KEY_SIZE; ++i) {
+        s_key[i] = rg->generate() & 0xFF;
+      }
+      
+      QString b64key(s_key.toBase64());
+      gboolean ok = secret_password_store_sync(cryptosettings_get_schema(),
+                                               SECRET_COLLECTION_DEFAULT,
+                                               "VPN settings encryption key",
+                                               qPrintable(b64key), nullptr,
+                                               &error, nullptr);
+      if (error != nullptr) {
+        logger.error() << "Key storage failed:" << error->message;
+        g_error_free(error);
+      }
+    }
+  }
+
+  if (s_key.length() == CRYPTO_SETTINGS_KEY_SIZE) {
+    memcpy(key, s_key.data(), CRYPTO_SETTINGS_KEY_SIZE);
+    return true;
+  }
+
+  logger.warning() << "Invalid key";
   return false;
 }
 
 // static
 CryptoSettings::Version CryptoSettings::getSupportedVersion() {
-  return CryptoSettings::NoEncryption;
+  return CryptoSettings::EncryptionChachaPolyV1;
```

Comment:
is there a way to check the libsecret status?

Reply:
Not in the official API as far as I can see, but we could check to see if the `org.freedesktop.secrets` name has been taken on the session D-Bus.

Reply:
yes, probably it's better. What do you think?

Reply:
I will add a check to ensure that `org.freedesktop.secrets` exists.